Ava UI: Implement built-in console window

merryhime opened the pull request:
<img width="912" alt="image" src="https://user-images.githubusercontent.com/8682882/211173408-0d772d0e-17d4-4030-bd0f-4a1ef4adee19.png">

Auto-scroll keeps scroll locked to end.
Open Logs open the logs folder for easy copying of logs.

Linked to Show Console configuration option:
<img width="266" alt="image" src="https://user-images.githubusercontent.com/8682882/211173412-cb15d409-436b-4771-b1eb-144171efd47f.png">

Note: I attempted to also capture stdout for mvk, but unfortunately the simple implementation using Console.SetOut/Console.SetError didn't work.

AcK77 commented on Ryujinx.Ava/UI/Windows/ConsoleWindow.axaml:
```diff
@@ -0,0 +1,65 @@
+                HorizontalAlignment="Center"
+                Name="AutoScrollCheckBox"
+                IsChecked="{Binding AutoScroll}"
+                Content="Auto-scroll" />
```
This should be added in Locale

AcK77 commented on Ryujinx.Ava/UI/Windows/ConsoleWindow.axaml:
```diff
@@ -0,0 +1,65 @@
+            <Button
+                HorizontalAlignment="Center"
+                Command="{ReflectionBinding OpenLogsFolder}"
+                Content="Open Logs" />
```
This should be added in Locale

AcK77 reviewed (changes requested):
About the design:
- The console title window should include the Ryujinx version as the original one.
- If `Auto-scroll` is checked, the scrollbar should be still active (even if it auto-scroll down, because if you stop the emulation then you can't scroll up to read the log without unchecking `Auto-scroll`.
- Maybe we should be able to select the "native" console or the Avalonia console as output in the settings or at least add `<OutputType>WinExe</OutputType>` in the csproj ?

I really like the idea, that could lead to more possibilities like some debug tools or anything related to debug.

merryhime commented:
Added in stderr redirection: stderr output from external code now goes through the logging system. Re-enable stdout output because not thinking of doing stdout redirection currently.

This allows for easier mvk/mesa debugging stderr will now show up in log files.

<img width="1147" alt="image" src="https://user-images.githubusercontent.com/8682882/211199213-42dcae6c-c290-436f-991c-59a302323f02.png">

merryhime commented:
@AcK77 

Good points.

* Added version to title as requested
* Enabled user scrolling when `Auto-scroll` is active. Originally the scrolling was just locked to always be at the bottom
* `WinExe` added
* Added `stderr` redirection
* I've also reenabled logging via stdout so running via the command line is always a fallback option if required

AcK77 commented on Ryujinx.Ava/UI/Models/InMemoryLogTarget.cs:
```diff
@@ -0,0 +1,78 @@
+
+        private static Color GetLogColor(LogLevel level) => level switch {
+            LogLevel.Info    => Colors.White,
+            LogLevel.Warning => Colors.Yellow,
+            LogLevel.Error   => Colors.Red,
+            LogLevel.Stub    => Colors.DarkGray,
+            LogLevel.Notice  => Colors.Cyan,
+            LogLevel.Trace   => Colors.DarkCyan,
```
```suggestion
            LogLevel.Warning => Color.FromRgb(235, 205, 85),
            LogLevel.Error   => Color.FromRgb(255, 95, 85),
            LogLevel.Stub    => Colors.DarkGray,
            LogLevel.Notice  => Color.FromRgb(2, 197, 229),
            LogLevel.Trace   => Color.FromRgb(200, 146, 245),
```

Just a little cosmetic suggestion, you can ignore it if you don't like.

![image](https://user-images.githubusercontent.com/4905390/211211583-a8516bae-87e3-452c-9948-d48e79126f1d.png)

gdkchan replied:
Personally I prefer the current colors, they are slightly brighter which makes it easier to read on a dark background.

AcK77 replied:
It's just following some of the Switch colors scheme

TSRBerry replied:
Suggested colors look great!
During testing I didn't notice any readability issues, but if they make messages harder to read, we should probably adjust them a bit.

AcK77 commented on Ryujinx.Ava/Program.cs:
```diff
@@ -1,5 +1,6 @@
 using Avalonia;
```
With "WinExe" added L126 `Console.Title = $"Ryujinx Console {Version}";` have to be removed because there is no Console and the emulator can't start!

merryhime replied:
Hah. Good catch.

GamerzHell9137 commented:
Idk if its known but there's 2 console windows now on Windows.

gdkchan commented on Ryujinx.Common/Logging/Logger.cs:
```diff
@@ -14,6 +15,8 @@ public static class Logger
 
         private static readonly List<ILogTarget> m_LogTargets;
 
+        private static readonly StdErrAdapter m_stdErrAdapter;
```
The naming here seems a bit weird. We usually don't use `m_` prefix, just `_`. And I think `m_` would not be correct here anyway since those are static fields? Those are usually prefixed with `s_`.

gdkchan commented on Ryujinx.Ava/UI/Models/InMemoryLogTarget.cs:
```diff
@@ -0,0 +1,78 @@
+            public string Text { get; set; }
+        }
+
+        private static Color GetLogColor(LogLevel level) => level switch {
```
```suggestion
        private static Color GetLogColor(LogLevel level)
        {
            return level switch
            {
```
Basically the same suggestion as the other PR.

gdkchan commented on Ryujinx.Ava/UI/Models/InMemoryLogTarget.cs:
```diff
@@ -0,0 +1,78 @@
+
+        public static InMemoryLogTarget Instance => _instance;
+
+        static InMemoryLogTarget()
```
Why does it need the empty static constructor?

gdkchan commented on Ryujinx.Ava/UI/Models/InMemoryLogTarget.cs:
```diff
@@ -0,0 +1,78 @@
+        private readonly ILogFormatter _formatter;
+        private readonly string _name;
+
+        private const int _maximumSize = 20000;
```
Since it's `const` it should be named `MaximumSize`.

gdkchan commented on Ryujinx.Ava/UI/Windows/ConsoleWindow.axaml.cs:
```diff
@@ -0,0 +1,63 @@
+
+        private void AutoScrollCheckBoxOnChecked(object sender, RoutedEventArgs e)
+        {
+            if (AutoScrollCheckBox.IsChecked == true)
```
```suggestion
            if (AutoScrollCheckBox.IsChecked)
```

merryhime replied:
IsChecked is `bool?`

gdkchan replied:
I see.

gdkchan commented on Ryujinx.Ava/UI/Windows/ConsoleWindow.axaml.cs:
```diff
@@ -0,0 +1,63 @@
+
+        private void ConsoleItemsControlOnLayoutUpdated(object sender, EventArgs e)
+        {
+            if (AutoScrollCheckBox.IsChecked == true && autoScrollChangeCount != changeCount)
```
```suggestion
            if (AutoScrollCheckBox.IsChecked && autoScrollChangeCount != changeCount)
```

gdkchan commented on Ryujinx.Ava/UI/Models/InMemoryLogTarget.cs:
```diff
@@ -0,0 +1,78 @@
+
+        private const int _maximumSize = 20000;
+
+        public ObservableCollection<Entry> Entries;
```
Can this be private? Ideally it should not be possible to modify internal state outside of the `InMemoryLogTarget` class.

merryhime replied:
This is exposed and bound to UI. Not sure how to achieve that otherwise.

gdkchan commented on Ryujinx.Ava/UI/Models/InMemoryLogTarget.cs:
```diff
@@ -0,0 +1,78 @@
+{
+    public class InMemoryLogTarget : ILogTarget
+    {
+        public class Entry
```
If the list is made private, I think this can become private too.

gdkchan commented on Ryujinx.Ava/UI/Models/InMemoryLogTarget.cs:
```diff
@@ -0,0 +1,78 @@
+    {
+        public class Entry
+        {
+            public Color Color { get; set; }
```
Unless there's a need to modify those values after instantiation, I think it would be better to remove the setter and initialize it on the constructor. Maybe you could make it a struct too to avoid the object allocation cost etc.

gdkchan commented on Ryujinx.Ava/UI/Models/InMemoryLogTarget.cs:
```diff
@@ -0,0 +1,78 @@
+
+        public ObservableCollection<Entry> Entries;
+
+        string ILogTarget.Name { get => _name; }
```
```suggestion
        string ILogTarget.Name => _name;
```
I think this is valid too?

gdkchan commented on Ryujinx.Common/Logging/LogLevel.cs:
```diff
@@ -10,6 +10,7 @@ public enum LogLevel
         Notice,
-        Trace
+        Trace,
+        StdErr,
```
Do we need a new log level for that? Can't we use the existing `Error` level?

merryhime replied:
I wasn't entirely sure what level to put it at; some of it is informational, some of it is errors.

TSRBerry commented:
I noticed the new console window can cause some performance issues:

When a lot of stuff gets logged really fast, the window can't keep up and freezes at some point. This also freezes the main window.
![image](https://user-images.githubusercontent.com/20988865/211233768-11525bcd-4bb9-4f42-a13d-77f61042bd3f.png)
I'm currently hearing the smash title song in the screenshot, but can't see anything because both windows are frozen.
Meanwhile the native console window still logs more messages.

Probably also related to very fast logging:
As long as the console window is open I drop some fps which causes noticeable audio stutters. 
![image](https://user-images.githubusercontent.com/20988865/211234030-96d0104b-6802-4d0f-901a-1b549c7d8789.png)

As soon as I close the console window the stutters are gone and my fps are back to normal:
![image](https://user-images.githubusercontent.com/20988865/211234267-fff5a32f-f639-4a99-bbd9-59a1b84104e1.png)

merryhime commented:
I suspect this is because ItemsPresenter doesn’t do virtualisation but will look into it.

merryhime commented:
I have also notice a disadvantage of this approach is that since this relies on the emulator being functional in order to have working output, sometimes things can be missed. For example, in a recent instance when the emulator threw a `System.InvalidOperationException`, this was shown on stdout and in file logs but not in this window, because the emulator had crashed.

MutantAura commented:
> I have also noticed a disadvantage of this approach is that since this relies on the emulator being functional in order to have working output, sometimes things can be missed. For example, in a recent instance when the emulator threw a `System.InvalidOperationException`, this was shown on stdout and in file logs but not in this window, because the emulator had crashed.

Is this a huge issue as long as it's in the file log? Realistically if the program has crashed then the console window would usually close shortly after anyway (at least on Windows this is the behaviour). It may also push more people to actually send us actual log files instead of console screenshots...

merryhime commented:
I personally feel there are reliability and inherent performance issues with a built in console window, so closing this.
